### PR TITLE
CIP-0030 | Upgrade to latest format & propose version / extension scheme

### PR DESCRIPTION
**Pull request by KtorZ**
This has been a recurring topic regarding extensions of CIP-0030.
  Given that this CIP is effectively active (hence the status change in
  this commit), we can't simply keep modifying it because it then
  becomes hard for DApps to keep track of what is supported by wallet
  providers and what isn't.

  In this commit, I henceforth propose an extension scheme based off
  CIPs. Wallet providers are expected to support the base API given by
  CIP-0030 and new extensions or modifications of that API can be
  brought in through extension CIPs. Upon initialization, Dapps can
  interrogate the wallet about what CIPs extensions it supports and
  process or fail accordingly based on whether the wallet provides the
  desired functionality.

### CODE REVIEW

**Comment by Quantumplation** [comment]
Rather than throwing an error if one or more of the features is unavailable, could we instead add a field to the returned object that is an array of the features it was able to turn on (or which the user selected to enable, for example)? this makes it the dApps choice on whether this is an error, or just degraded functionality, without having to call enable repeatedly with progressively fewer features.

For example, I may request a wallet with notifications support, but still be able to operate just fine without it.

Or, I may ask to enable a "prompt less signing" extension that allows more rapid "day-trader" like experiences, but the user declines.  The app will still work, and I don't want to binary search through the space of possibilities.

You could use this same mechanism to decide on the *version* of an extension, much like an HTTP content type negotiation.

**Comment by KtorZ** [comment]
> @Quantumplation could we instead add a field to the returned object that is an array of the features it was able to turn on

That's a good idea; what about adding that to the error response actually?

**Comment by Quantumplation** [comment]
@KtorZ well, the idea is, currently `enable` returns an object that has all the methods we might use to start interacting with the wallet; if we throw an error, we're interrupting the return of that object.

The idea behind my suggestion is to *still return the same object* from enable; if I absolutely require some feature that isn't in `(await cardano.{wallet}.enable()).extensions`, then I can treat that as an error and show the user some kind of explanatory dialog.  But, if it was just a nice to have, I can just change the behavior of my application to omit that missing feature.

If, instead, we threw an error that just described "here are the features we could turn on if you wanted, and here are those we couldn't", then I have to call enable again with a restricted set, which is going to (potentially) prompt the user again, etc.

**Comment by KtorZ** [comment]
> @Quantumplation which is going to (potentially) prompt the user again

This is a valid UX point indeed.

**Comment by Quantumplation** [comment]
Riffing on this a bit more, maybe a bit more structure would be useful as well.  To be a bit more concrete, here's what I'm suggesting:

 - `enable(extensions)` doesn't return an error;
 - instead, a new field is added to the returned object, called `extensions`
 - this field is an object; the keys on this object is the union of extensions the wallet knows about, and the ones that were requested
 - If a particular extension wasn't requested, but is supported, the value is `null`; this allows the dApp to know which features it *could* leverage, if it needed (and, for example, show/hide different flows)
 - If a particular extension was requested, and enabled, the value here is an object with all methods relevant to the particular extension
 - If a particular extension was requested and couldn't be enabled for some reason, the value is an error describing why it wasn't enabled (Incompatible, NotSupported, UserDeclined, etc.)

If people don't like the "null" value to discover features (totally get it, I'm iffy on it too), then I'd suggest adding a list of potentially supported extensions to the initial object (the one at `window.cardano.{wallet}`) instead.

**Comment by KtorZ** [comment]
I went for something half-way while you were writing your comment. So, no error but a new endpoint to get the list of extensions actually enabled by the wallet. 

> @Quantumplation  I'd suggest adding a list of potentially supported extensions to the initial object (the one at window.cardano.{wallet}) instead.

I think that's also easy to do and provide benefits indeed. I am afraid that something more complicated will rapidly be a pain to implement for wallet providers.

**Comment by Scitz0** [comment]
It's a good thing to get some structure for extensions to CIP-30. Though the currently suggested string array of extensions to enable limits you from adding additional data in the enable call that might be needed. Taking CIP-0062 as an example it now has its own enable call (that this addition is supposed to replace?) and takes an additional VotingPurpose[] in the enable call. 

I would like to extend Extension from being a string array to an object array containing:
```js
{
  "cip": <number>,
  "ver": <number>,
  ...
}
```

- `cip` is the number of the CIP extension to CIP-0030 to request access for
- `ver` is the version number of the extension that the wallet support
- `...` object can contain any additional fields as specified by the CIP extension. like purpose array for CIP-0062

The same Extension object as above would be returned when calling `supportedExtensions()`. This would allow for example a wallet that support CIP-0062 extension to return cip, ver, and the supported purposes.

Does this make sense? 
I must admit that I was thinking it through as I wrote the reply 😄

**Review by rphair — APPROVED** [review]
This proposal looks fine as far as I will currently be able to understand it; I know some discussion will want to happen (above my level) but I'm approving this now based on:
- formatting improvements & compatibility with new CIP-0001 guidelines
- adding history of CIP's acceptance & implementation
- method `cardano.{walletName}.isEnabled()` answering long standing question about versioning / feature support including issues like https://github.com/cardano-foundation/CIPs/pull/443

**Comment by Scitz0** [comment]
> @KtorZ Thus, there's in principle no need for version numbers because CIPs do not generally have versions. They have new CIPs instead.

Ok, I can accept the workflow of removing versioning and always requiring a new CIP for extension/changes to existing CIP. 

But how would you propose that additional data in enable call should be handled? 
Like requesting access to a specific feature within a CIP (that CIP-0062 currently utilizes).

**Comment by Scitz0** [comment]
We should then remove apiVersion from CIP-30 as well, as it makes no sense if a CIP, once merged, is final and all changes/extensions should be handled by new CIPs.

**Comment by Quantumplation** [comment]
@Scitz0 
> Like requesting access to a specific feature within a CIP (that CIP-0062 currently utilizes).

One way to handle this would be to leave that negotiation to the extension as a separate method;

For example, you call `enable("CIP-0062")`, and then call `.enablePurposes(...)`, a method provided by the extension;

I'm not super familiar with CIP-0062, so I don't know if this would re-trigger a new user prompt, but if so, that's a slight downside UX-wise.

**Comment by Scitz0** [comment]
Internal discussions for how to deal with the proposed workflow have resulted in I'm being ok with the PR in its current state. The only thing would be why `Extension` is formatted as a string and not just a number representing the CIP number? But I'm ok with it either way.

Just as an FYI, being really more of a CIP-62 thing. I think we will instead of accepting the purpose as an input to enable specific purposes, have CIP-0062 _Extension_ enable call (according to this PR) enable all supported governance purposes for an origin and instead add a getSupportedPurposes() where the wallet can specify the supported purposes. The apiVersion in CIP-62 will also be removed to only allow minor changes, like purpose definitions. 

Is this in line with the desired outcome of this PR?

File: CIP-0030/README.md [hunk 1]
```diff
@@ -158,57 +187,65 @@ type TxSignError = {
...
 
-
-
-## Initial API
+### Initial API
 
 In order to initiate communication from webpages to a user's Cardano wallet, the wallet must provide the following javascript API to the webpage. A shared, namespaced `cardano` object must be injected into the page if it did not exist already. Each wallet implementing this standard must then create a field in this object with a name unique to each wallet containing a `wallet` object with the following methods. The API is split into two stages to maintain the user's privacy, as the user will have to consent to `cardano.{walletName}.enable()` in order for the dApp to read any information pertaining to the user's wallet with `{walletName}` corresponding to the wallet's namespaced name of its choice.
 
-### cardano.{walletName}.enable(): Promise\<API>
+#### cardano.{walletName}.enable(extensions: Extension[] = []): Promise\<API>
```
Comment (ehanoc) [inline CIP-0030/README.md:194]: Should `Extension` just be a string? I takes perhaps away the ability to pass more information upon requesting a new extension, which can be useful for the user to make a decision to enable it or not. For example in #296 `cardano.{walletName}.governance.enable(purpose: VotingPurpose[]): Promise\<API>`

Comment (KtorZ) [reply CIP-0030/README.md:194]: Not sure to follow you? What would be the structure of that extension string?

Comment (Quantumplation) [reply CIP-0030/README.md:194]: @KtorZ I think ehanoc is asking about/suggesting something like

```
cardano.{walletName}.enable([{ cip: "CIP-0062", parameters: { purpose: ["catalyst", "sundae"] } }]);
```

I like the simplicity of the string approach, since we could balloon forever on complexity here (enable by patterns, versioning, fallbacks, etc.)

That being said, I can see the UX risk if the alternative is to prompt the user twice: once to enable an extension, and another to configure that extension.

In this case, perhaps it would be sufficient for each voting purpose to be a separate CIP that you could request, but it's only one step further where that starts to break down; Imagine a hypothetical CIP that allows wallets to sign transactions without a prompt (for instant trade experiences):

```
cardano.walletName.enable({ cip: "CIP-XXXX", spendingLimit: { coins: 100000000, assets: {...} } })
```

**Comment by alessandrokonrad** [comment]
What would be the major difference between the [Experimental API](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030#experimental-api) and extensions? Isn't the goal anyway that most wallets support a certain extension (if popular/used a lot). If every wallet just adds whatever extensions they like it's going to be a mess for dApp developers anyway. So over time it's desired that most wallets support popular extensions, but then they should be at the base, because it's not an extension anymore then really. If extensions don't see adoption over time then they are likely dropped again, so they are "experimental".

**Comment by teddyjfpender** [comment]
> cardano.{walletName}.enable(extensions: Extension[] = []): Promise<API>
I really like this, particularly as the set of possible extensions is large. 

I wonder though, from the wallet user's perspective how they might wish to enable a subset of the extensions a dApp requested at initialisation; and maintain a whitelist of the enabled (or not to be enabled) extensions per dApp. Likewise from the dApp's perspective they may make a hard requirement on a subset of the extensions requested at initialisation. 

Allowing the dApp to specify a required subset of extensions and giving the wallet user the ability to selectively enable or disable extensions requested at initialisation for different dApps could provide greater flexibility and control for both parties. It could also improve the UX of a wallet, as the wallet user can choose to only enable extensions that they trust and are comfortable with.

**Comment by KtorZ** [comment]
> @alessandrokonrad: What would be the major difference between the [Experimental API](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030#experimental-api) and extensions?

These are two orthogonal elements. The experimental API is (purposely) loosely specified. It's a way for wallets to provide specific features that aren't actually specified as CIP extensions. Over time, I wish we would deprecate the experimental API altogether because it creates inconsistencies in the ecosystem. It still bothers me a lot when I see a list of "supported wallets" on a DApp whereas the entire point of CIP-0030 was to remove the need for doing adhoc integration. A DApp should support *any* wallet that implement CIP-0030. 

Yet how did we actually get there? 

-> There are actually many divergent implementations of CIP-0030.

Which is why I am proposing to freeze the current API (+ this extension mechanism) and require that all wallet (willing to follow this standard) implement at least the base API and advertise what other features they support through this extension mechanism.

This is truly about features negotiation. So it becomes much more transparent for dapps and wallets alike. Instead of integrating against a wallet, a DApp can integrate against a set of interfaces. Any wallet that provide those interfaces should work. 

I didn't remove the experimental API because I think there's still a (narrow) use case for it. If only during development of cutting-edge features not yet fully specified; but I would encourage any wallet providers to document their interfaces as CIPs rather than shoving stuff in the experimental API so we can grow the whole ecosytem.

**Comment by KtorZ** [comment]
> @teddyjfpender from the wallet user's perspective how they might wish to enable a subset of the extensions a dApp requested at initialisation

> @teddyjfpender Allowing the dApp to specify a required subset of extensions and giving the wallet user the ability to selectively enable or disable extensions

I believe this to be out of scope of the current specification. CIP-0030 is a contract between dapps softwares and wallet softwares; users aren't personas of this particular subsystem.

What I mean to say is that, there's nothing in the specification that constraints how wallets may want to handle what extensions they make available. So, if a wallet provider wants to leave the choice to their users through some settings dashboard they're welcomed to do so; this specification doesn't prevent it.

The most naive implementation of this proposal would be for wallets to simply implement and enable all extensions they support every time. Some may want to do something more fine grained.

Note that the requested extensions that the dapp specifies on initialisation is mainly to cope with possibly conflicting extensions and give a hint to the wallet to which one they should enable if they support multiple conflicting extensions.


Imagine two extensions CIP-1234 and CIP-5678. The former provides a way to evaluate a transaction execution units by providing a method 'evaluateTx' in the API that returns some cpu and memory budgets. The other is an extension that performs all the phase-1 validations on a transaction (kind of dry submission) through a method also called 'evaluateTx' (!).


Both extensions are mutually incompatible because they are both bound to a same endpoint with different semantics. This is where a DApp could indicate which one it wants enabled during initialisation. In practice, well obviously do our best to avoid extensions that conflicts like this, unless it is a desired outcome (eg a new version of a particular extension that provides new functionalities but with a non backward compatible interface).

**Comment by alessandrokonrad** [comment]
> It still bothers me a lot when I see a list of "supported wallets" on a DApp whereas the entire point of CIP-0030 was to remove the need for doing adhoc integration. A DApp should support _any_ wallet that implement CIP-0030.

But isn't that almost unavoidable? When I look at JavaScript builtin functions you also see a list of browsers supporting that builtin or not. Extensions do not solve that problem, they just indicate better what wallets are supported by a certain dApp. 
When a dApp relies on a certain feature of an extension and not all wallets support that extension then again you have that list of "supported wallets". I'm not opposed to extensions, I'm just failing to see this. Is the idea to break down bigger compatibility issues into smaller ones?
 
> Which is why I am proposing to freeze the current API (+ this extension mechanism) and require that all wallet (willing to follow this standard) implement at least the base API and advertise what other features they support through this extension mechanism.

When you say freeze, you mean "forever", or under what circumstances can the base API be modified?

**Comment by KtorZ** [comment]
> @alessandrokonrad But isn't that almost unavoidable? When I look at JavaScript builtin functions you also see a list of browsers supporting that builtin or not. Extensions do not solve that problem

And I am not saying they do! 

> @alessandrokonrad they just indicate better what wallets are supported by a certain dApp.

Precisely! This is the entire goal here. Have a less ad-hoc mechanism for wallets and dapps to advertise and negotiate what features they support / need. In principle, this could now happen entirely programmatically. When writing your dApp integration, you could simply look for any available wallet, filter out those who don't support your mandatory extensions, and automatically propose them to the user while leaving the others out. As soon as one wallet brings support for a required extension, no work is needed from your dApp -> it automatically supports it. 

Of course, there will always be some little details about one wallet implementing something slightly differently than others. But that is something we should resolve at the specification-level. Improve the CIPs so that they lead to unambiguous implementations. 

> @alessandrokonrad When you say freeze, you mean "forever", or under what circumstances can the base API be modified?

Freeze (mostly) forever yes. Have any changes be brought be new CIPs. But it's very possible that a CIP extension would override an existing method. Say that tomorrow, we change the interface to `signTx`, this could be written as an extension. And wallets would enable it on-demand from DApps. 

There are cases where we might want to make exceptional changes to the base CIP-0030 API; but we will have to be very careful about that to avoid breaking backward-compatibility. This is why I am keen on @Quantumplation's idea (which he proposed on another discussion thread) to have extensions be plain objects with a single key "cip" for now. For example:

```
{ "cip": 62 }
```

This allows to possibly introduce later some extra arguments to extensions without breaking all existing interfaces.

**Comment by gitmachtl** [comment]
i find it a bit confusing to have totally different cip numbers that are basically all variations or extensions of the original one. isn't it time to group them under a main cip number and do extentions/variations on that via a minor naming scheme? like 802.a, 802.n, 802.ac, 802.ax ...? they are all describing wifi specifications/protocols/frequencies etc, but we know they all have to do with network stuff. its just an idea, but maybe we should think about it to group such things.

**Comment by KtorZ** [comment]
@gitmachtl that's a good point but kind of orthogonal to CIP-0030. It'd have appreciated the feedback when we reworked CIP-0001 and discussed numbering there as well 😶 ... 

There was many debates about how to number CIPs but all-in-all, we concluded that what is really needed is _some_ number that allows to identify CIPs and refer to them easily. We could come up with something more elaborate but every proposal so far had trade-offs we didn't want to introduce. For lack of better proposal, we chose to keep it simple

**Comment by Ryun1** [comment]
This is a good proposal IMO.

The only issue I have is that it removes the potential for flexibility in extensions' `enable` call, which has been highlighted above. Another outcome of this is that extension specific `initial API` functions will not be possible.

I think it may be useful for some extra prose to be added around guidance for extensions. Questions, I have:
- can extensions depend on other extensions? - can extensions have extensions?
- should extensions follow a specific format?
- can extensions add their own error codes? - can they overwrite error codes?
- should extensions have a clear size? - what is too small to be an extension what is too large?
- should extensions have a clear focus (i.e. governance) or can they be generic?

Probably ridged boundaries around scope and size are inappropriate, with extensions just being considered against community consensus of the problems they solve.

**Comment by Quantumplation** [comment]
Another extension to the idea behind this PR could be to make `CIP-0030` one of the CIP's to enable; an empty array is assumed to be `CIP-0030`.

This opens up the possibility of a future CIP-0030 defining a different "base" interface, if we want to make changes to any of those "base" methods.

For example, if we wanted to change the signature of signData or signTxs, or rework how things work entirely (pub-sub streaming communication with the wallet, or something).

I'm more than happy with this proposal without it, just turning it over in my head to see if there's anything we're going to regret not considering 2 years from now 😅

**Comment by rphair** [comment]
> @gitmachtl: extentions/variations on that via a minor naming scheme? like 802.11a, 802.11n, 802.11ac, 802.11ax
> @KtorZ: It'd have appreciated the feedback when we reworked CIP-0001 and discussed numbering there as well ...

Just recalling that we _did_ consider this, though it was someone else that brought up the idea last time.  I responded at the time against the idea but what I said was not a comprehensive treatment, nor did it necessarily reflect the opinions of all editors, nor did what I write determine how CIP-0001 was finalised... it was just some specific reasons to "keep [the names] simple" as we might also have been generally & instinctively inclined to do.
- https://github.com/cardano-foundation/CIPs/pull/366#issuecomment-1309650355
- https://github.com/cardano-foundation/CIPs/pull/366#issuecomment-1310468382

**Comment by Quantumplation** [comment]
@thaddeusdiamond the problem is, there is no *one* API; this proposal is about how a dApp and a CIP negotiate features, not (just) a version.

That is, there may be methods that the wallet doesn't want to support; or that it wants to gather user input before enabling.  It's akin to the permissions an app requests from your phone, for example. Some phones don't have a camera, and sometimes the user wants to disallow camera access but allow other features.

This proposal is about allowing dApps and wallets to navigate that process, with little knowledge of each other beyond the shared touch point of a CIP for standardization.

The *governance* side of things, i.e the discussion, consensus, and adoption of these is provided by the process outlined in CIP-0001; the individual CIPs themselves document the specific extension itself.

You bring up a good point that perhaps it's worth collecting these decisions into a separate format for documentation and ease of use, but I think that's orthogonal to this proposal, i.e. feature negotiation.

**Comment by KtorZ** [comment]
CIPs aren't about governance though. We've repeated that countless times -- and even write it quite clearly on CIP-0001 I believe now. Though I understand the confusion.

The CIPs process is a mechanism to reach agreement on technical solutions. CIPs are the technical solutions. Whether a party then wants to implement and support a solution is up to them. That's even more true of the wallet APIs. If tomorrow, a new CIP is accepted, it doesn't automatically create a commitment to every wallet implementors out there. It can't.

CIPs are a catalog of solutions. And the suggestion here is to allow both ends (dapps and wallets) to inform each other about which parts of the catalog they need and support. I'd argue that DApps developers must know about CIPs -- at least to one that are relevant to them. At the moment, this feature negotiation is very cumbersome because everyone has their own subset of features they support. And we end up in situations where each integration requires efforts. 

Collecting CIP-0030 related extensions into one website (as well as a compatibility matrix of which wallets support which) sounds like a good idea. It is akin to what https://www.cardano-caniuse.io/ is trying to achieve.

**Comment by thaddeusdiamond** [comment]
I love cardano-caniuse.io and as a dApp developer use it regularly.  Let me try another analogy.

Multiple web browsers have their own implementations of various CSS features, but they all align to one standard.  The three-phase process is roughly something like:

1. The Web 3 Consortium puts out a proposal for a new CSS module (e.g., [border-radius](https://www.w3.org/TR/2002/WD-css3-border-20021107/))
2. Browsers adopt it experimentally with varying prefixes (sidenote: this is kind of a nightmare for developers but it is what it is, see "Mozilla Equivalent" and "WebKit Equivalent" [here](https://www.the-art-of-web.com/css/border-radius/)).
3. Aggregators and learning sites like StackOverflow and w3schools.com create simplified tutorials for web developers consumption.

I would argue that dApp developers are equivalent to web devs, who never ever ever read the W3C papers.  Those are the equivalent of CIPs.  Part 2 should be Eternl/Nami/Flint's unique documentation.  Part 3 is cardano-caniuse.com

So my proposed solution would be for each addition to the Wallet API go through it's own CIP, completely detached from any dApp developer's knowledge of the API.

That make sense?

**Comment by KtorZ** [comment]
> @thaddeusdiamond this is kind of a nightmare for developers but it is what it is

How can you say that, and then not see the value of what we are proposing here 😅 ... ? We don't have to replicate a scheme that is _a nightmare_ for people to use. It doesn't have to be "what it is".

> @thaddeusdiamond So my proposed solution would be for each addition to the Wallet API go through it's own CIP

Now I've lost you. This is exactly what is being proposed here. With the addition of a mechanism for wallets to clearly indicates which addition they implement (so that there's no reliance on some external service to tell you empirically whether a wallet supports your needed feature). 

> @thaddeusdiamond completely detached from any dApp developer's knowledge of the API

I think all is really needed to get you onboard here is a clear API reference on a website that fully documents the API, and the known extensions. Take a step back, and imagine that instead of calling the extensions `CIP-0093` and `CIP-0123`, we call them `v2` and `v3`. And we just have an API reference that lists the features of v1, v2 and v3. So as a developer, you know which one you need for your use case. 

There's not much difference here, except that we don't have a pure sequence, but instead, a core foundation (CIP-0030) and composable extensions that can be added as needed.

**Comment by KtorZ** [comment]
@gitmachtl yes indeed. That's what I am suggesting with this "website" idea. Something on top of CIPs that is specific to the wallet API and its extensions. And if people want to replace CIP names with fruits or I-don't-know-what, then so be it.

File: CIP-0030/README.md [hunk 2]
```diff
@@ -1,72 +1,92 @@
...
-## Version
-
-The API specified in this document will count as version 0.1.0 for version-checking purposes below.
-
-## Data Types
-
-### Address
+#### Address
 
 A string representing an address in either bech32 format, or hex-encoded bytes. All return types containing `Address` must return the bech32 format, but must accept either format for inputs.
```
Comment (mkazlauskas) [inline CIP-0030/README.md:50]: Might be out of scope of this PR, but if we do this:
> Which is why I am proposing to freeze the current API

Then we should do this:
```suggestion
A string representing an address in either bech32 format, or hex-encoded bytes. All return types containing `Address` must return the hex-encoded bytes format, but must accept either format for inputs.
```
Because it's how the cip is implemented in practice

Comment (KtorZ) [reply CIP-0030/README.md:50]: Sounds reasonable to me.

**Comment by KtorZ** [comment]
Updated with recent feedback ☝️

**Comment by Quantumplation** [comment]
@KtorZ did that latest force push include some accidental changes to the template / CIP-0001?

**Comment by rphair** [comment]
@Quantumplation maybe what look like accidental changes are from these 2 already merged commits which affect incidental material?
- https://github.com/cardano-foundation/CIPs/pull/450
- https://github.com/cardano-foundation/CIPs/pull/452

... as far as I know these commits appearing in the diff for this branch are a consequence of doing a force-push rather than re-basing this branch on the `master` which already had those commits merged.

**Comment by KtorZ** [comment]
Nothing's accidental. Everything's planned. I think Github is just drunk here in showing the diff.

**Comment by thaddeusdiamond** [comment]
Moved my discussion with @KtorZ into a review to track better.

**Comment by Mercurial** [comment]
@KtorZ any issues with using https://www.cardano-caniuse.io maybe it can just be the start ?

File: CIP-0030/README.md [hunk 3]
```diff
@@ -1,72 +1,93 @@
 ---
```
[the remark this answers is not in the document]
Comment (KtorZ) [reply CIP-0030/README.md:1]: In order to have sequencing, you need to have consensus between wallet providers on the features that gets added and this is simply impossible.
Not all wallets may 

(a) be okay with a particular API extensions even though it's legit and backed by many, 

(b) have time / resources to keep up with every new API change. 

A more modular versioning à la carte solves this as people can implement what they truly care about.

**Comment by KtorZ** [comment]
No issue in using it. Rather, about keeping it up to date. In the end, it provides an idea but doesn't necessarily reflect reality. Embedding the version negotiation in the API itself solves this problem as wallet and DApp can dynamically figure out where they stand. 

I'd imagine cardano-caniuse to still be useful to (a) provide an overview of the extensions available and what wallet implement them and (b) still indicate information about experimental apis of wallets.

File: CIP-0030/README.md [hunk 4]
```diff
@@ -76,10 +97,17 @@ type Paginate = {|
 ```
 Used to specify optional pagination for some API calls. Limits results to {limit} each page, and uses a 0-indexing {page} to refer to which of those pages of {limit} items each. dApps should be aware that if a wallet is modified between paginated calls that this will change the pagination, e.g. some results skipped or showing up multiple times but otherwise the wallet must respect the pagination order.
 
+#### Extension
+
+An extension is an object with a single field `"cip"` that describe a CIP number extending the API (as a plain integer, without padding). For example:
```
[the remark this answers is not in the document]
Comment (KtorZ) [reply CIP-0030/README.md:102]: And you don't have to if you do not want. You can always ignore it and do with whatever the wallet provides you. Yet, the mechanism exists for those who seek this granularity.

[the remark this answers is not in the document]
Comment (Quantumplation) [reply CIP-0030/README.md:102]: An alternative would be to require each CIP to unambiguously define a slug that describes the API like "notifications v1" but this is even less discoverable or consistent.  In practice the cardano community has rallied around referring to things by CIP number anyway ("is the wallet CIP-30 compliant?", "Is this a CIP-68 compatible NFT", etc.)

Even if you're unaware of the details or history of the CIP, you're aware of the name, which gives you a shared frame for googling, discussing, finding libraries for, etc.

File: CIP-0030/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Quantumplation) [reply CIP-0030/README.md:1]: Think of this less like versions of *an* API, and more like permissions you're requesting: some phones have cameras, NFC readers, flexible screens, etc, but others dont; so as a dApp developer, you're negotiating with the wallet for which features you *can* leverage, which will never be homogeneous.

Some wallets will be unable or unwilling to implement notifications, or fast-signing, or network/account switching, etc. But *will* be willing and able to implement other CIPs.

The fact that this supports gracefully evolving those apis as well is an additional bonus.

File: CIP-0030/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply CIP-0030/README.md:102]: You'd still have to look up the cip(s) for what's supported in which versions. Don't see a difference in terms of dApp developer effort.

We can (and will) build libraries on top of dApp connector cip(s), which can abstract away the cip(s) and have opinionated APIs. Having said that, I'm not sure we should use the current revision cip-0030 as an immutable base. For example, all wallets have to implement `getCollateral` to be compliant, but it should be deprecated at some point, as using 'collateral return' is a superior way of handling the collateral (any utxo can be used as a collateral).

@KtorZ how about drafting a new CIP that would define the dapp connector without **any** wallet functionality, but with extensions, where cip-0030 wallet API is one of the extensions?

[the remark this answers is not in the document]
Comment (KtorZ) [reply CIP-0030/README.md:102]: @mkazlauskas that would have been the ideal approach. But given that CIP-0030 is already somewhat implemented by everyone already. I am arguably just trying to make it slightly easier to manage all conversations down-the-line w.r.t adding stuff to CIP-0030. From a CIP perspective, it's been a bit hard to manage evolutions of CIP-0030 precisely because adding anything to it pushes work onto everyone and it's hard to find consensus. 

Having said that, I wouldn't be against _trimming down_ the current CIP-0030 to a bare minimum, and move core functionality into separate CIPs. Yet I have not really the bandwidth to write and champion those other CIPs (plural because I think the current API should rather be expressed as separate points).

[the remark this answers is not in the document]
Comment (Quantumplation) [reply CIP-0030/README.md:102]: CIP-30 certainly isn't an immutable base; a future CIP could replace/modify it, and then dApps would signify they're ready to upgrade by passing in that CIP number, and wallets would gradually support it.

Eventually we may get to a state where in practice nobody passes an empty array, and no wallets supports the very old CIP-30 and fail if you do pass an empty array.

It "feels" unclean because you'll have some APIs hanging around for a long time, but in practice that's what backwards compatibility, graceful evolution, and long term support *means*.

Another way to think about it:
 
CIP-30 is just another extension you're enabling, but it's assumed to be passed by default for backwards compatibility.

[the remark this answers is not in the document]
Comment (KtorZ) [reply CIP-0030/README.md:102]: Yup

[the remark this answers is not in the document]
Comment (Ryun1) [reply CIP-0030/README.md:102]: > how about drafting a new CIP that would define the dapp connector without any wallet functionality, but with extensions, where cip-0030 wallet API is one of the extensions?

I am keen to take a swing at this and see what people think 😎

[the remark this answers is not in the document]
Comment (rphair) [reply CIP-0030/README.md:102]: @Ryun1 @KtorZ I think that will save _so_ much grief in the long run... imagine if, in its own early days, developers had made an inventory of the few protocols which would "always" be in the Internet: and how unworkable that would seem as a baseline today 🧐

[the remark this answers is not in the document]
Comment (rphair) [reply CIP-0030/README.md:102]: Resolved by https://github.com/cardano-foundation/CIPs/pull/462 (@thaddeusdiamond)?  At least with respect to the feature baseline (leaving aside the nomenclature issues)...?

[the remark this answers is not in the document]
Comment (thaddeusdiamond) [reply CIP-0030/README.md:102]: > Resolved by #462 (@thaddeusdiamond)? At least with respect to the feature baseline (leaving aside the nomenclature issues)...?

It is resolved in terms of this issue, but I'd actually rather that linked specification use the slugs as someone mentioned below `collateral-experimental`.  Naming is hard enough, linking them to numbers adds yet another layer of obscurity.

It makes me think of the old CS saying: "There are only two hard problems in computer-science: floating points, naming, and off-by-one errors".

[the remark this answers is not in the document]
Comment (rphair) [reply CIP-0030/README.md:102]: @thaddeusdiamond then I'm resolving this thread & can you or @Quantumplation please post your suggestion about the naming convention in #462? (I don't see it there) cc: @Ryun1

[the remark this answers is not in the document]
Comment (rphair) [reply CIP-0030/README.md:102]: @thaddeusdiamond then I'm resolving this thread & can you or @Quantumplation please post your suggestion about the naming convention in #462? (I don't see it there) cc: @Ryun1

File: CIP-0030/README.md [hunk 5]
```diff
@@ -158,57 +186,82 @@ type TxSignError = {
...
 
-
-
-## Initial API
+### Initial API
 
 In order to initiate communication from webpages to a user's Cardano wallet, the wallet must provide the following javascript API to the webpage. A shared, namespaced `cardano` object must be injected into the page if it did not exist already. Each wallet implementing this standard must then create a field in this object with a name unique to each wallet containing a `wallet` object with the following methods. The API is split into two stages to maintain the user's privacy, as the user will have to consent to `cardano.{walletName}.enable()` in order for the dApp to read any information pertaining to the user's wallet with `{walletName}` corresponding to the wallet's namespaced name of its choice.
 
-### cardano.{walletName}.enable(): Promise\<API>
+#### cardano.{walletName}.enable(extensions: Extension[] = []): Promise\<API>
```
Comment (SebastienGllmt) [inline CIP-0030/README.md:193]: I strongly recommend 

`request: {extensions: Extension[]}`

instead. Structural arguments are easier to change or extend later than positional arguments

File: CIP-0030/README.md [hunk 6]
```diff
@@ -1,72 +1,93 @@
...
 
-# Specification
-
-## Version
-
-The API specified in this document will count as version 0.1.0 for version-checking purposes below.
+#### Address
 
-## Data Types
+A string representing an address in either bech32 format, or hex-encoded bytes. All return types containing `Address` must return the hex-encoded bytes format, but must accept either format for inputs.
```
Comment (SebastienGllmt) [inline CIP-0030/README.md:51]: Why did this PR change the address format for return types from bech32 to hex-encoded? Seems unrelated to this PR

Comment (rphair) [reply CIP-0030/README.md:51]: according to how it's "implemented in practice": https://github.com/cardano-foundation/CIPs/pull/446#discussion_r1102723847

Comment (KtorZ) [reply CIP-0030/README.md:51]: Because it was indicated that this is how most wallets have implemented it today. And the goal was to reflect on the current reality.

File: CIP-0030/README.md [hunk 7]
```diff
@@ -158,57 +186,82 @@ type TxSignError = {
...
+
+DApps are expected to use this endpoint to perform an initial handshake and ensure that the wallet supports all their required functionalities. Note that it's possible for two extensions to be mutually incompatible (because they provide two conflicting features). While we may try to avoid this as much as possible while designing CIPs, it is also the responsability of wallet providers to assess whether they can support a given combination of extensions, or not. Hence wallets aren't expected to fail should they not recognize or not support a particular combination of extensions. Instead, they should decide what they enable and reflect their choice in the `cardano.{walletName}.extensions` field of the Full API. As a result, dApps may fail and inform their users or may use a different, less-efficient, strategy to cope with a lack of functionality.
```
Comment (Ryun1) [inline CIP-0030/README.md:201]: `cardano.{walletName}.extensions`

Is referenced here, but it is never defined in the specification.

Comment (Ryun1) [reply CIP-0030/README.md:201]: ill make a PR to fix.

Comment (Ryun1) [reply CIP-0030/README.md:201]: Actually, I believe to just be a typo, this should reference `api.getExtensions()` and not `cardano.{walletName}.extensions`.
